circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -0,0 +1,1242 @@
+        old_wallet: WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_wallet: WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        external_transfer: ExternalTransferVar,
+        update_timestamp: Variable,
+        cs: &mut CS,
+    ) {
+        // External transfer must have binary direction
+        ConstrainBinaryGadget::constrain_binary(external_transfer.direction, cs);
+
+        // Validate updates to the orders within the wallet
+        Self::validate_order_updates(&old_wallet, &new_wallet, update_timestamp, cs);
+
+        // Validate updates to the balances within the wallet
+        Self::validate_balance_updates(&old_wallet, &new_wallet, external_transfer, cs);
+    }
+
+    // ------------
+    // | Balances |
+    // ------------
+
+    /// Validates the balance updates in the wallet
+    fn validate_balance_updates<CS: RandomizableConstraintSystem>(
+        old_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        external_transfer: ExternalTransferVar,
+        cs: &mut CS,
+    ) {
+        // Ensure that all mints in the updated balances are unique
+        Self::constrain_unique_balance_mints(new_wallet, cs);
+        // Validate that the external transfer has been correctly applied
+        Self::validate_external_transfer(old_wallet, new_wallet, external_transfer, cs);
+    }
+
+    /// Validates the application of the external transfer to the balance state
+    pub(crate) fn validate_external_transfer<CS: RandomizableConstraintSystem>(
+        old_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        new_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES, LinearCombination>,
+        external_transfer: ExternalTransferVar,
+        cs: &mut CS,
+    ) {
+        // Zero out the external transfer amount if its mint is zero
+        let external_transfer_zero = EqZeroGadget::eq_zero(external_transfer.mint, cs);
+        let external_transfer_not_zero = NotGate::not(external_transfer_zero, cs);
+        let (_, _, transfer_amount) =
+            cs.multiply(external_transfer.amount.into(), external_transfer_not_zero);
+
+        // The term added to any balance matching the external transfer mint
+        let external_transfer_term = CondSelectGadget::select(
+            -transfer_amount,
+            transfer_amount.into(),
+            external_transfer.direction.into(),
+            cs,
+        );
+
+        // Stores a boolean indicating whether the external transfer appeared in the new balances
+        // it must be applied to at least one balance
+        let mut external_transfer_applied: LinearCombination = Variable::Zero().into();
+
+        // Check the balance state transition for each balance
+        for (old_balance, new_balance) in old_wallet.balances.iter().zip(new_wallet.balances.iter())
```

Comment:
I'm a little confused by the structure of this iteration.
it seems to rely on a proper ordering of balances to be zipped, i.e. any pair in the zip must be either:
1. (`mint_i`, `mint_i`)
2. (`zero_mint`, `mint_i`)
3. (`mint_i`, `zero_mint`)
but a simple off-by-one could throw off the whole circuit, i.e.:
```
(`zero_mint`, `mint_i`)
(`mint_i`, `zero_mint`)
```
how do we know the iterators are properly sorted / have the right number of zero mints/zero balances in the right places? is this handled in the `Iterator` impl for `wallet.balances`?

Reply:
The prover should not shuffle the balances in between old and new wallet, see the test cases below. So the inputs to the circuit should come sorted. The validity checks on the mints and balances effectively enforce that all but one balance is updated, and that this balance update is consistent with the external transfer. So as long as the witness is not shuffled we're good